### PR TITLE
Implemented task-api

### PR DESCRIPTION
**Pull request by wachterjohannes**
Add task entity and controller.

Basic PR: https://github.com/sulu/sulu/pull/3037

### CODE REVIEW

**Comment by wachterjohannes** [comment]
@danrot

File: src/Sulu/Bundle/AutomationBundle/Controller/TaskController.php [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+     *
+     * @return Response
+     */
+    public function getAction($id)
+    {
+        $manager = $this->getTaskManager();
+        $task = $manager->findById($id);
+
+        if (!$task) {
+            throw new NotFoundHttpException();
```
Comment (danrot) [inline src/Sulu/Bundle/AutomationBundle/Controller/TaskController.php]: Actually we changed that to using the `FOSRestBundle` exception functionality for that. I would throw a `TaskNotFoundException` in the `findById` method, and configure the corresponding http error code in the FOSRestBundle configuration.

File: src/Sulu/Bundle/AutomationBundle/Controller/TaskController.php [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    /**
+     * Returns task-manager.
+     *
+     * @return TaskManagerInterface
+     */
+    private function getTaskManager()
+    {
+        return $this->get('sulu_automation.tasks.manager');
```
Comment (danrot) [inline src/Sulu/Bundle/AutomationBundle/Controller/TaskController.php:188]: Open question: Should we implement controller as services now?

File: src/Sulu/Bundle/AutomationBundle/Events/CancelTrait.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    /**
+     * Cancel Event.
+     *
+     * @param bool $canceled
+     *
+     * @return $this
+     */
+    public function cancel($canceled = true)
```
Comment (danrot) [inline src/Sulu/Bundle/AutomationBundle/Events/CancelTrait.php]: What is the parameter for?

File: src/Sulu/Bundle/AutomationBundle/Resources/config/services.xml [hunk 4]
```diff
@@ -9,5 +9,23 @@
...
         </service>
+
+        <service id="sulu_automation.tasks.manager" class="Sulu\Bundle\AutomationBundle\Tasks\Manager\TaskManager">
+            <argument type="service" id="sulu.repository.task"/>
+            <argument type="service" id="event_dispatcher"/>
+        </service>
+
+        <!-- FIXME hack to enable doctrine object constructor -->
+        <!--
+          could be fixed in serializer by changing JMS\SerializerBundle\DependencyInjection\Compiler\DoctrinePass:49
```
Comment (danrot) [inline src/Sulu/Bundle/AutomationBundle/Resources/config/services.xml:20]: What's the state about this?

File: src/Sulu/Bundle/AutomationBundle/Resources/config/doctrine/Task.orm.xml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    <entity name="Sulu\Bundle\AutomationBundle\Entity\Task" table="au_task">
+        <id name="id" type="integer" column="id">
+            <generator strategy="AUTO"/>
+        </id>
+
+        <field name="entityClass" type="string" column="entityClass"/>
+        <field name="entityId" type="string" column="entityId"/>
+
+        <field name="taskName" type="string" column="taskName"/>
+        <field name="schedule" type="datetime" column="schedule"/>
```
Comment (danrot) [inline src/Sulu/Bundle/AutomationBundle/Resources/config/doctrine/Task.orm.xml:14]: How do we handle recurring events?

File: src/Sulu/Bundle/AutomationBundle/Tasks/Model/TaskInterface.php [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\AutomationBundle\Tasks\Model;
+
+use Sulu\Component\Persistence\Model\AuditableInterface;
+
+/**
+ * Interface TaskInterface.
```
Comment (danrot) [inline src/Sulu/Bundle/AutomationBundle/Tasks/Model/TaskInterface.php]: Better PHPDoc.

**Review by danrot — CHANGES_REQUESTED** [review]

File: src/Sulu/Bundle/AutomationBundle/Controller/TaskController.php (hunk 2 above)
Comment (wachterjohannes) [reply src/Sulu/Bundle/AutomationBundle/Controller/TaskController.php:188]: @chirimoya

File: src/Sulu/Bundle/AutomationBundle/Resources/config/services.xml (hunk 4 above)
Comment (wachterjohannes) [reply src/Sulu/Bundle/AutomationBundle/Resources/config/services.xml:20]: we would have to update jms-serializer first. before that it makes less sense to fix it there.

File: src/Sulu/Bundle/AutomationBundle/Resources/config/doctrine/Task.orm.xml (hunk 5 above)
Comment (wachterjohannes) [reply src/Sulu/Bundle/AutomationBundle/Resources/config/doctrine/Task.orm.xml:14]: we have currently no recuring events. but then we would have to add some additional fields here (like cron-string or other stuff).

File: src/Sulu/Bundle/AutomationBundle/Events/CancelTrait.php (hunk 3 above)
Comment (wachterjohannes) [reply src/Sulu/Bundle/AutomationBundle/Events/CancelTrait.php]: hmm i think its not needed because we cannot "undo" it.

Comment (wachterjohannes) [reply src/Sulu/Bundle/AutomationBundle/Events/CancelTrait.php]: i will remove it

File: src/Sulu/Bundle/AutomationBundle/Controller/TaskController.php (hunk 2 above)
Comment (wachterjohannes) [reply src/Sulu/Bundle/AutomationBundle/Controller/TaskController.php:188]: i have shortly discussed this with @chirimoya and we will go wi this way. maybe refactoring in a later PR

**Comment by wachterjohannes** [comment]
@danrot fixed

File: src/Sulu/Bundle/AutomationBundle/Controller/TaskController.php [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+     *
+     * @throws TaskNotFoundException
+     */
+    public function getAction($id)
+    {
+        $manager = $this->getTaskManager();
+        $task = $manager->findById($id);
+
+        if (!$task) {
+            throw new TaskNotFoundException($id);
```
Comment (danrot) [inline src/Sulu/Bundle/AutomationBundle/Controller/TaskController.php]: I would have moved that to the manager. Then the exception will also be thrown in other use cases.

**Review by danrot — CHANGES_REQUESTED** [review]

**Pull request merged**